llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1752,6 +1752,76 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
   }
 }
 
+/// Hash the underlying data of a VPSingleDefRecipe pointer, instead of hashing
+/// the pointer itself.
+namespace {
+struct VPCSEDenseMapInfo : public DenseMapInfo<VPSingleDefRecipe *> {
+  static bool isSentinel(const VPSingleDefRecipe *Def) {
+    return Def == getEmptyKey() || Def == getTombstoneKey();
+  }
+
+  static bool canHandle(const VPSingleDefRecipe *Def) {
+    return isa<VPInstruction, VPWidenRecipe, VPWidenCastRecipe,
+               VPWidenSelectRecipe, VPHistogramRecipe, VPPartialReductionRecipe,
+               VPReplicateRecipe, VPWidenIntrinsicRecipe>(Def);
```

Comment:
I think we should also check `VPRecipeBase::mayHaveSideEffects()` too

Reply:
Checking for this regresses the current result, and I think VPlan's mayHaveSideEffects() may be too conservative.

Reply:
you at least need to skip recipes that may read memory I thin. Otherwise you may reuse the value from an earlier load, which gets modified by a store in between.

if it just has side effects without reading state, that should be fine as we don’t remove the side effect causing recipe.

Probably the only relevant case would be stores, which don’t produce a value to replace.

Reply:
Yeah, that should be left for removeDeadRecipes or at least the removal needs to be guarded by a check for side-effects

Reply:
We definitely need test-coverage for cases like multiple replicating stores to the same address with different values.

Reply:
More to the point of what canHandle here means: could we add them to the map: yes, but we could not remove them after replacing their uses